### ISSUE TITLE
Messages are stuck in rsyslog queue when using RELP delivery

### ISSUE THREAD

**Issue by jay7x** [rsyslog/rsyslog#1782]
I have following action configured in ruleset on API server. This ruleset receive messages from PHP process (`fwrite()` to `stream_socket_client()`-connected socket) via `imuxsock` input.

```
  action(type="omrelp" name="XXX_sock_relay"
    target="10.X.X.X"
    port="2510"
    tls="off"
    tls.compression="off"
    template="XXX_relay_msg"
    action.reportSuspensionContinuation="on"
    action.resumeRetryCount="-1"
    action.resumeInterval="10"
    queue.type="fixedArray"
    queue.size="250000"
    queue.dequeueBatchSize="1024"
    queue.workerThreads="4"
    queue.workerThreadMinimumMessages="50000"
    queue.spoolDirectory="/shop/logs/rsyslog"
    queue.fileName="XXX_sock"
    queue.maxFileSize="16m"
    queue.maxDiskSpace="2G"
    queue.highWaterMark="200000"
    queue.lowWaterMark="100000"
    queue.checkpointInterval="30000"
    queue.saveOnShutdown="on"
    queue.timeoutEnqueue="1"
  )
```

Few days ago I've started to investigate why this queue `XXX_sock_relay` always have 1000-4000 messages in it when other API's queues have < 10 messages usually.

I've realised this queue on every API server have few old messages stuck in it. Every few mins those messages are delivered to log collector. But those messages seems are not marked as delivered after! So they are remaining in queue to be delivered again and again. Delivery attempt happens every 10s.

I have saved tcpdump for this but it contains private info I don't like to put in public. So if you're interested in investigation of this case please let's contact directly.

I've switched partially to TCP delivery (omfwd/imptcp). There is no issues at all. So it's exactly about `omrelp`/`imrelp`.

CentOS 7 x86_64
rsyslog 8.27
librelp 1.2.14

**Comment by mostolog** [comment]
I would bet we have seen similar issues in the past. Did you try deleting your queues? A fresh start could help solving the issue, although possible error may not be fixed.

**Comment by jay7x** [comment]
New messages will appears stuck in queue in about 24h after fresh start.

**Comment by deoren** [comment]
I too am seeing messages stuck in the queue (v8.29.0 from the official adiscon Ubuntu PPA). I end up restarting rsyslog and after the systemd timeout is exceeded, the rsyslog daemon is killed. When rsyslog starts back up it complains about the corrupt queue and all messages that were held are lost.

After a time messages begin queuing again and only a restart on the client gets rsyslog going again.

I'm using a DA queue connected to a omrelp action. Nothing relevant appears to be logged on the receiver (also Ubuntu 16.04, also v8.29.0 from PPA).

From the client with debug logging enabled (debug on demand):

```
9090.427044446:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relp engine is dispatching frame with command 'rsp'
9090.427047229:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: in rsp command handler, txnr 1, code 200, text 'OK'
9090.427143165:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: DEL sess 0x7f7558007550 unacked 1535, sessState 2
9090.427149145:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: processing server offer 'commands'
9090.427151668:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: SetEnableCmd in syslog cmd state: 4
9090.427153911:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: SetEnableCmd out syslog cmd state: 4, iRet 0
9090.427156022:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: enabled command 'syslog'
9090.427158125:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: processing server offer 'relp_software'
9090.427160217:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: processing server offer 'relp_version'
9090.427162897:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: in destructor: sendbuf 0x7f75580b9d20
9090.427165317:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: iRet after relpSessRcvData 0
9090.427167561:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpSessWaitState returns 0
9090.427169655:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: end relpSessConnect, iRet 0
9090.427172049:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relp session 0x7f7558007550 reestablished, now resending 1535 unacked frames
9090.427216903:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: resending frame '2 syslog 14059 <189>2017-09-27T08:55:36.760929-05:00 redminebox redmine-prod @cee:{ "metadata": { "filename": "\/opt\/redmine\/log\/production.log" }, "tag": "redmine-prod", "msg": "REAL MESSAGE REMOVED" }

'
9090.427255851:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpTcpSend: sock 8, lenbuf 14075, send returned 14075 [errno 11]
9090.427259371:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: sendbuf NOT added to unacked list
9090.427261895:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: resending frame '3 syslog 226 <189>2017-09-27T08:55:52.548235-05:00 redminebox redmine-prod @cee:{ "metadata": { "filename": "\/opt\/redmine\/log\/production.log" }, "tag": "redmine-prod", "msg": "Completed 200 OK in 29ms (Views: 2.7ms | ActiveRecord: 7.9ms)" }

'
9090.427265349:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpTcpSend: sock 8, lenbuf 240, send returned 240 [errno 11]
9090.427267199:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: sendbuf NOT added to unacked list
9090.427269628:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: resending frame '4 syslog 285 <189>2017-09-27T08:55:52.550444-05:00 redminebox redmine-prod @cee:{ "metadata": { "filename": "\/opt\/redmine\/log\/production.log" }, "tag": "redmine-prod", "msg": "REAL MESSAGE REMOVED" }
```

Sample of messages from `/var/log/rsyslog.log` :

```
2017-10-15T13:13:09.450430-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:13:09.450708-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:07.438448-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:07.438823-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:27.450556-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:27.450774-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:49.522824-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:49.523363-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:57.477119-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:59.534695-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:15:59.534920-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:16:07.482431-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:16:07.482694-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:16:09.546364-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:16:37.502905-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:16:49.566536-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:16:49.566741-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:37.530619-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:37.531068-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:39.602643-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:39.602840-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:47.542374-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:47.542639-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:49.614637-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
2017-10-15T13:17:49.614806-05:00 redminebox rsyslogd: action 'ForwardToSawmill1' resumed (module 'omrelp') [v8.29.0 try http://www.rsyslog.com/e/2359 ]
```

@rgerhards @davidelang Is this a known issue?

I've tried both with setting the `target` value to the FQDN (original setting) and a bare IP Address (current setting, same problem).

**Comment by davidelang** [comment]
you  need to configure systemd to give rsyslog more time to shut down to avoid 
the corrupt queue files (there is a utility to make index files for the queue 
files so that you can recover the data that is in them)

There are some relp issues that have been found very recently, test the current 
github version (or wait for the next release, due shortly) to see if that 
addresses your problem.

**Comment by rgerhards** [comment]
Note: 8.29.0 has some additional queue hardening, but of course things can always go wrong if kill -9 is emitted. Upon start rsyslog should complain about (most) queue errors.

**Comment by davidelang** [comment]
since we have the utility that can recreate the .qi files, how hard would it be 
to have a startup option for rsyslog that, if the queue files are corrupt uses 
similar logic to replay the contents of the files? (possibly with some logic 
like mv queue queue.bad and then read logs from the queue.bad and either send 
them or let them build up in new queue files)

**Comment by rgerhards** [comment]
@davidelang that'a a good idea, but I think it should go into its own tracker.

**Comment by deoren** [comment]
> @davidelang need to configure systemd to give rsyslog more time to shut down to avoid the corrupt queue files 

I currently have systemd configured to give rsyslog 240 seconds to shut down cleanly, up from the default for Ubuntu 16.04 of 90 seconds. What setting do you use or would you suggest in this case?

> @davidelang There are some relp issues that have been found very recently, test the current github version (or wait for the next release, due shortly) to see if that addresses your problem.

I see that the v8.30.0 milestone is closed. I plan to update all of our Ubuntu boxes to v8.30.0 as soon as it hits the official PPA.

Re the error messages, what situation usually results in errors like these?

```
9090.427255851:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpTcpSend: sock 8, lenbuf 14075, send returned 14075 [errno 11]
9090.427259371:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: sendbuf NOT added to unacked list
```

When you see those errors logged, is it usually an issue with the sending rsyslog instance or a remote receiver instance? I assume the former since our other nodes are currently OK (around 40-50 of them).

**Comment by davidelang** [comment]
On Mon, 16 Oct 2017, Deoren Moor wrote:

>> @davidelang need to configure systemd to give rsyslog more time to shut down 
>> to avoid the corrupt queue files
>
> I currently have systemd configured to give rsyslog 240 seconds to shut down 
> cleanly, up from the default for Ubuntu 16.04 of 90 seconds. What setting do 
> you use or would you suggest in this case?

I would try shutting it down manually and see how long it takes to finish. The 
more data it has to flush to queue files, the longer it will take.

>> @davidelang There are some relp issues that have been found very recently, test the current github version (or wait for the next release, due shortly) to see if that addresses your problem.
>
> I see that the v8.30.0 milestone is closed. I plan to update all of our Ubuntu boxes to v8.30.0 as soon as it hits the official PPA.
>
> Re the error messages, what situation usually results in errors like these?
>
> ```
> 9090.427255851:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpTcpSend: sock 8, lenbuf 14075, send returned 14075 [errno 11]
> 9090.427259371:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: sendbuf NOT added to unacked list
> ```
>
> When you see those errors logged, is it usually an issue with the sending rsyslog instance or a remote receiver instance? I assume the former since our other nodes are currently OK (around 40-50 of them).

I don't know, rainer may have a comment on them.

**Comment by deoren** [comment]
> @davidelang: I would try shutting it down manually and see how long it takes to finish. The
more data it has to flush to queue files, the longer it will take.

Sorry to be dense, but how should I stop it? I usually call either `service rsyslog stop` or `systemctl stop rsyslog`, but in either case for an Ubuntu 16.04 box it would use the systemd timeout for the unit. I've already set it to use 240 seconds for the timeout. In the past, that timeout has been reached whether I try to reboot the box or whether I ask systemd to stop it.

**Comment by davidelang** [comment]
kill -TERM pid will tell it to shutdown gracefully (all the other stuff is a 
bunch of wrappers around this)

**Comment by deoren** [comment]
Thanks @davidelang. I did that:

`kill -TERM $(cat /var/run/rsyslogd.pid)`

and 15 minutes later the daemon was still running. After about 24 minutes however, the `/var/log/rsyslog.log` file was updated:

```
2017-10-16T14:32:20.892998-05:00 czer rsyslogd:  [origin software="rsyslogd" swVersion="8.29.0" x-pid="825" x-info="http://www.rsyslog.com"] exiting on signal 15.
2017-10-16T14:56:10.048577-05:00 czer rsyslogd: rsyslogd's groupid changed to 103
2017-10-16T14:56:10.048867-05:00 czer rsyslogd: rsyslogd's userid changed to 101
2017-10-16T14:56:10.048936-05:00 czer rsyslogd: queue 'strm 0x25cd1a0', file '/var/spool/rsyslog/q_ForwardToSawmill1.00000884' opened for non-append write, but already contains 584351 bytes  [v8.29.0 try http://www.rsyslog.com/e/0 ]
2017-10-16T14:56:10.049007-05:00 czer rsyslogd:  [origin software="rsyslogd" swVersion="8.29.0" x-pid="5672" x-info="http://www.rsyslog.com"] start
```

Messages from 2017-10-16T19:32:20.892Z forward were relayed properly. Everything older, was not.

`root@redminebox:/var/log# journalctl -u rsyslog.service | grep -E 'Start|Stop'`

```
Oct 15 10:28:27 redminebox systemd[1]: Stopping System Logging Service...
Oct 15 10:32:27 redminebox systemd[1]: Stopped System Logging Service.
Oct 15 10:32:40 redminebox  systemd[1]: Starting System Logging Service...
Oct 15 10:32:41 redminebox systemd[1]: Started System Logging Service.
Oct 16 14:56:09 redminebox systemd[1]: Starting System Logging Service...
Oct 16 14:56:10 redminebox systemd[1]: Started System Logging Service.
```
Evidently systemd started the service up again once it noticed that it had stopped.

**Comment by deoren** [comment]
> I too am seeing messages stuck in the queue (v8.29.0 from the official adiscon Ubuntu PPA). I end up restarting rsyslog and after the systemd timeout is exceeded, the rsyslog daemon is killed. When rsyslog starts back up it complains about the corrupt queue and all messages that were held are lost.

I misspoke earlier when I mentioned that the DA queue was cleared upon restart of the rsyslog daemon. I missed it when manually reviewing the details, but I ended up spotting it when checking the details after my last post:

`grep -i ForwardToSawmill1 /var/log/rsyslog-pstats.log | grep 'size=' | tail -n 2`

```
Mon Oct 16 16:10:14 2017: ForwardToSawmill1 queue[DA]: origin=core.queue size=1239008 enqueued=0 full=0 discarded.full=0 discarded.nf=0 maxqsize=0
Mon Oct 16 16:10:14 2017: ForwardToSawmill1 queue: origin=core.queue size=0 enqueued=26516 full=0 discarded.full=0 discarded.nf=0 maxqsize=3829
```
From what I can tell from occasionally checking the output from the chain grep calls, the queue count is slowly dropping. I've not yet determined if the messages are arriving at the receiver.

**Comment by deoren** [comment]
@jay7x, as you noted, switching to `omfwd` appears to allow messages to flow. 

Are you by chance working with a system that generates larger than usual log messages? In my case, the one system having more trouble than any other is where I'm using imfile to read in some very busy flat-files. The message is then wrapped in a JSON payload for the receiver to extract and save locally before forwarding on to two downstream rsyslog receivers.

It was when I switched to omfwd that I saw messages like this this one on the receiver:

```
2017-10-18T11:58:42.178969-05:00 sawmill1 rsyslogd: imptcp imptcp: message received is at least 315 byte larger than max msg size; message will be split starting at: " <4>2017-10-16T16:56:08.108693-0"  [v8.30.0]
```

**Comment by deoren** [comment]
@rgerhards @davidelang  As I just mentioned on the mailing list thread, it appears that explicitly setting `MaxDataSize="128k" ` for the imrelp input on our main receiver "resolves" the problem for the sending node. Messages that appeared to be infinitely "stuck", now flow as expected.

The configuration now:

```
global (
    maxMessageSize="128k"
    action.reportSuspension="on"
    action.reportSuspensionContinuation="on"
    workDirectory="/var/spool/rsyslog"
    processInternalMessages="off"
    parser.permitSlashInProgramName="on"
)

# more config settings here ...

input(
    type="imrelp"
    port="2514"
    KeepAlive="on"
    ruleset="receiver-rsyslog-JSON-payload-parsing"
    MaxDataSize="128k"
)
```

Setting just `maxMessageSize="128k"` on its own does not appear to be sufficient to resolve the problem.

From the imrelp documentation:

> MaxDataSize <size_nbr>
> Default is the global message size
>
> Sets the max message size (in bytes) that can be received. Any messages above this size will be rejected causing the relp client to reconnect and retry. 

Speaking from ignorance, it appears that messages like the following indicate a reconnect/retry loop as noted in the documentation, but I don't see anything there that clearly (at least to me) explains that the client is being rejected due to an overly large message size. Is there support for logging that reason either in the debug or the standard log output?

```
9090.427044446:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relp engine is dispatching frame with command 'rsp'
9090.427047229:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: in rsp command handler, txnr 1, code 200, text 'OK'
9090.427143165:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: DEL sess 0x7f7558007550 unacked 1535, sessState 2
9090.427149145:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: processing server offer 'commands'
9090.427151668:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: SetEnableCmd in syslog cmd state: 4
9090.427153911:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: SetEnableCmd out syslog cmd state: 4, iRet 0
9090.427156022:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: enabled command 'syslog'
9090.427158125:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: processing server offer 'relp_software'
9090.427160217:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: processing server offer 'relp_version'
9090.427162897:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: in destructor: sendbuf 0x7f75580b9d20
9090.427165317:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: iRet after relpSessRcvData 0
9090.427167561:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpSessWaitState returns 0
9090.427169655:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: end relpSessConnect, iRet 0
9090.427172049:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relp session 0x7f7558007550 reestablished, now resending 1535 unacked frames
9090.427216903:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: resending frame '2 syslog 14059 <189>2017-09-27T08:55:36.760929-05:00 redminebox redmine-prod @cee:{ "metadata": { "filename": "\/opt\/redmine\/log\/production.log" }, "tag": "redmine-prod", "msg": "REAL MESSAGE REMOVED" }

'
9090.427255851:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpTcpSend: sock 8, lenbuf 14075, send returned 14075 [errno 11]
9090.427259371:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: sendbuf NOT added to unacked list
9090.427261895:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: resending frame '3 syslog 226 <189>2017-09-27T08:55:52.548235-05:00 redminebox redmine-prod @cee:{ "metadata": { "filename": "\/opt\/redmine\/log\/production.log" }, "tag": "redmine-prod", "msg": "Completed 200 OK in 29ms (Views: 2.7ms | ActiveRecord: 7.9ms)" }

'
9090.427265349:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: relpTcpSend: sock 8, lenbuf 240, send returned 240 [errno 11]
9090.427267199:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: sendbuf NOT added to unacked list
9090.427269628:ForwardToSawmill1 queue[DA]:Reg/w0: omrelp.c: resending frame '4 syslog 285 <189>2017-09-27T08:55:52.550444-05:00 redminebox redmine-prod @cee:{ "metadata": { "filename": "\/opt\/redmine\/log\/production.log" }, "tag": "redmine-prod", "msg": "REAL MESSAGE REMOVED" }
```

**Comment by deoren** [comment]
@rgerhards @davidelang As a follow-up to my last posting, is there a way to see in the debug logs what value imrelp is using for imrelp's `MaxDataSize` parameter if don't explicitly set it, and instead rely on imrelp to use the global() `maxMessageSize` value?

On that note, when `maxMessageSize` is inherited by an input, is that value written out to the logs for troubleshooting purposes?

**Comment by rgerhards** [comment]
see also https://github.com/rsyslog/rsyslog/issues/1741

**Comment by deoren** [comment]
> @rgerhards @davidelang As a follow-up to my last posting, is there a way to see in the debug logs what value imrelp is using for imrelp's MaxDataSize parameter if don't explicitly set it, and instead rely on imrelp to use the global() maxMessageSize value?
> 
> On that note, when maxMessageSize is inherited by an input, is that value written out to the logs for troubleshooting purposes?

The latest `v8-devel` build contains the changes introduced in 69661aac9082ce68daafdfe97715917f96b3c93e to display both of those values in the debug log. I upgraded our test receiver and then enabled debug logging. I grepped the debug log and got back these hits:

`root@sawmill3:~# grep -iE 'imrelp: max data|max message size' /var/log/rsyslog-debug-all.log`
```
7813.599060984:main thread    : imrelp.c: imrelp: max data size 131072
7813.599139377:main thread    : imrelp.c: imrelp: max data size 8096
7813.599215819:main thread    : imrelp.c: imrelp: max data size 8096
7813.599292279:main thread    : imrelp.c: imrelp: max data size 8096
7813.599367773:main thread    : imrelp.c: imrelp: max data size 8096
7813.607027063:main thread    : rsyslogd.c: max message size: 131072
```

I then disabled where I had explicitly set the `MaxDataSize` imrelp parameter to match the value I had chosen for `global(maxMessageSize)`. I deleted the debug log and restarted rsyslog. Now I see different results:

`root@sawmill3:~# grep -iE 'imrelp: max data|max message size' /var/log/rsyslog-debug-all.log`
```
7897.228013779:main thread    : imrelp.c: imrelp: max data size 8096
7897.228095527:main thread    : imrelp.c: imrelp: max data size 8096
7897.228178757:main thread    : imrelp.c: imrelp: max data size 8096
7897.228260197:main thread    : imrelp.c: imrelp: max data size 8096
7897.228339995:main thread    : imrelp.c: imrelp: max data size 8096
7897.235886677:main thread    : rsyslogd.c: max message size: 131072
```

It does not appear that imrelp is using the global `maxMessageSize` value as intended.

I'm about to draft a simplified test config and attempt to replicate those results in a separate Ubuntu 16.04 test VM. I'll post the results and the conf file once that is finished.

**Comment by deoren** [comment]
1. Enable v8-devel branch 
    `apt-add-repository ppa:adiscon/v8-devel`
1. Update rsyslog and rsyslog-relp packages
    `sudo apt-get install rsyslog rsyslog-relp`
1. Drop in this configuration: [rsyslog-i1782-receiver.conf](https://github.com/deoren/rsyslog-examples/blob/master/github_issues/rsyslog-i1782-receiver.conf)
1. Restart rsyslog

`root@ubuntu-virtual-machine:/var/log# grep -iE 'imrelp: max data|max message size' /var/log/rsyslog-debug-full.log`
```
0690.539898435:main thread    : imrelp.c: imrelp: max data size 8096
0690.542792875:main thread    : rsyslogd.c: max message size: 131072
```

**Comment by deoren** [comment]
@rgerhards I know you have a ton of items on the TODO list, but if/when you circle back to this and are ready to have someone test changes and I'll be happy to do so.

**Comment by deoren** [comment]
> happy to have someone to check this out when it is ready. TBH I am just "recovering from the past two weeks" and this is not on top of the list. I would expect that I can't get to it in 8.34 timeframe. Next target now is the imfile refactoring (I just finished the meeting where we decided that).

ACK.

**Comment by rgerhards** [comment]
I am finally trying to tackle this issue, together with some other librelp related work. Will keep you updated here. Now first need to review how librelp internally handles the message buffers on send and receive.

**Comment by rgerhards** [comment]
Question: assuming the RELP client (sender) is told to send an oversize message: what shall it do? I would tend to

* log a warning diagnostic (maybe error)
* truncate

If we just error out, this can (and probably will) trigger loops in the user app (here: omrelp). After all, if it did not check for the size, it is unclear if it will handle a size-related error code correctly. 

Note: At this point in processing, splitting the message is no option. That's something to consider in the upper layers (imfile e.g.).

**Comment by rgerhards** [comment]
I should clarify that I talk about librelp, which is the first thing to be fixed. And, yes, I guess I found the right spots.

**Comment by jay7x** [comment]
Is it possible to negotiate max message size when `omrelp` is connecting to `imrelp`? Then we may configure behaviour on connection stage.

I think we should log warning on connection stage (once per connection attempt) and truncate message then.

**Comment by deoren** [comment]
> @rgerhards: Question: assuming the RELP client (sender) is told to send an oversize message: what shall it do? I would tend to
> 
> log a warning diagnostic (maybe error)
> truncate

From the [current docs](https://www.rsyslog.com/doc/v8-stable/configuration/modules/imrelp.html):

> Sets the max message size (in bytes) that can be received. Any messages above this size will be rejected causing the relp client to reconnect and retry.

My 2 cents:

1. log the rejection (#2190), note the specific reason (mention the max size allowed and also the size of the message that was rejected) and consider that an error. 
1. log a separate message for the truncation of the message. 
    - I'm not sure if that would best be handled as a warning or error, but I would personally not consider it anywhere near the same level as imfile's, "you did not specify an input file, so I will do nothing" situation, so perhaps an error for that as well. 

If you only generate one message, then I'd suggest having the one be logged as an error.

> If we just error out, this can (and probably will) trigger loops in the user app (here: omrelp). After all, if it did not check for the size, it is unclear if it will handle a size-related error code correctly.

Maybe this is what you're saying, but it appears that a loop was occurring as of https://github.com/rsyslog/rsyslog/issues/1782#issuecomment-352027757; it appeared that the remote imrelp receiver was rejecting messages coming in that were larger than the default 8096 max data size value (`maxMessageSize` was not being honored) and the sender just kept retrying. Without any log messages explicitly noting the rejection, I found myself in a trial & error situation until what I believe was some advice from @davidelang that led me to setting both the `maxMessageSize` and `MaxDataSize` parameters (instead of just setting `maxMessageSize`).

For my specific needs, having the logged rejections would probably be enough to detect and resolve the problem for desired message traffic (likely by increasing the `maxMessageSize` value), but having an optional ability to enable truncation as a fall-back would be really helpful.

**Comment by davidelang** [comment]
On Thu, 12 Apr 2018, Rainer Gerhards wrote:

> just as an update: a root cause of the problem is that imfile did not properly 
> obey maxmessage size - and maybe some other parts of the core engine. I also 
> need to think on a method how we can really enforcee maxmessage size - we 
> should not permit any message larger than the configured max to be able to 
> float through the engine.

The problem is what to do when you hit a size limit. Especially if it's a matter 
of the output of a template being too large as a result of including a JSON tree 
in it.

As an admin, I want the ability to figure out that the message is too large to 
send remotely, adn write it to a local file and modify the message to send a 
smaller version to the remote destination

**Comment by PascalWithopf** [comment]
@jay7x @deoren @davidelang 
I was assigned to this issue and read through it and related issues, but it is not really clear to me what exactly the issue is here.

**To sum up**
- The issue was created because messages get stuck in queues when using relp.
- Because everything is working when using omfwd and imtcp instead, it is assumed that the problem lies in relp.
- It was discovered that the imrelp parameter maxDataSize doesn't use the global parameter maxMessageSize as default value.
- A discussion was started about the behaviour when messages are sent that are too long.
- #1741 was mentioned. Even when maxDataSize is higher than 128k, messages higher than 128k are not being delivered.
- #2633 was mentioned. When messages longer than maxMessageSize are send they will never get acknowledged and stopping rsyslog will end in a segmentation fault.

**My questions**
- Is maxMessageSize still not setting the default for maxDataSize? or was this fixed.
- What is the current problem in this issue?

**Comment by deoren** [comment]
> @PascalWithopf: My questions
>
> Is maxMessageSize still not setting the default for maxDataSize? or was this fixed.

Thanks to #2215, this is now logged in the debug output. As of https://github.com/rsyslog/rsyslog/issues/1782#issuecomment-352027757 the two were not in sync as shown in that output. I've not tested since then, but can do so if needed. https://github.com/rsyslog/rsyslog/issues/1782#issuecomment-352041870 shows the steps I took to replicate the results from the last test that I performed (showing that setting the value for `maxMessageSize` did not also update the default for `maxDataSize`).

> @PascalWithopf: What is the current problem in this issue?

I can't speak for others, but when I last tested, larger than usual messages from omrelp clients were stuck in the queue until I adjusted the imrelp `maxDataSize` parameter on the receiver to accommodate the non-default message size (greater than 8k).

**Comment by deoren** [comment]
> @PascalWithopf: I was assigned to this issue

I forgot to say it in the last post I made, but many thanks for taking this on.

**Comment by deoren** [comment]
@PascalWithopf I repeated the earlier steps from https://github.com/rsyslog/rsyslog/issues/1782#issuecomment-352041870 a few minutes ago and then grepped the debug log as shown below:

`root@ubuntu-1404-virtual-machine:~# grep -iE 'imrelp: max data|max message size|version' /var/log/rsyslog-debug-full.log`

Matched entries from the debug log:

```
0744.471385268:main thread    : imrelp.c: imrelp: max data size 8096
0744.478656490:main thread    : iminternal.c: signaling new internal message via SIGTTOU: ' [origin software="rsyslogd" swVersion="8.35.0.c3fb6e07e270" x-pid="2926" x-info="http://www.rsyslog.com"] start'
0745.485912722:main thread    : rsyslogd.c: max message size: 131072
```

**Comment by PascalWithopf** [comment]
@deoren thanks for the explanation. I'm working with @rgerhards in librelp on a parameter to specify how messages that are to long should be handled and other issues.

**Comment by rgerhards** [comment]
fixed in #2656
requires librelp 1.2.16
